File: SW1/csharp/2/UserService.cs
namespace CourseManagementSystem;

public sealed class UserService
{
    private readonly SQLiteSchoolRepository _repository;
    private readonly UserFactory _userFactory = new();
    private int _nextUserId;

    public Dictionary<string, User> Users { get; } = new(StringComparer.OrdinalIgnoreCase);

    public UserService(SQLiteSchoolRepository repository)
    {
        _repository = repository;
        Users = _repository.LoadUsers();
        _nextUserId = Users.Values
            .Select(user => int.TryParse(user.UserId, out var userId) ? userId : 0)
            .DefaultIfEmpty(0)
            .Max();
    }

    public User RegisterUser(string username, string password, string email, string role)
    {
        if (Users.ContainsKey(username))
        {
            throw new InvalidOperationException("Username already exists");
        }

        _nextUserId++;
        var userId = _nextUserId.ToString();
        var user = _userFactory.CreateUser(username, password, email, userId, role);
        Users[username] = user;
        _repository.AddUser(user);
        return user;
    }

    public User LoginUser(string username, string password)
    {
        if (!Users.TryGetValue(username, out var user))
        {
            throw new InvalidOperationException("User does not exist");
        }

        if (user.Password != password)
        {
            throw new InvalidOperationException("Incorrect password");
        }

        return user;
    }

    public User? FindUser(string username)
    {
        return Users.TryGetValue(username, out var user) ? user : null;
    }

    public bool DeleteUser(string username)
    {
        if (!Users.Remove(username))
        {
            throw new InvalidOperationException("User does not exist");
        }

        _repository.DeleteUser(username);
        return true;
    }

    public List<User> FindUsersByRole(string role)
    {
        return Users.Values
            .Where(user => string.Equals(user.Role, role, StringComparison.OrdinalIgnoreCase))
            .ToList();
    }
}

File: SW1/2/src/service/User_service.py
from UserManager.user_factory import UserFactory
from models.user import User


class UserService:
    def __init__(self):
        self.user_factory = UserFactory()
        self.users: dict[str, User] = {}
        self.next_user_id = 0

    def register_user(self, username, password, email, role) -> User:
        if username in self.users:
            raise ValueError("Username already exists")
        self.next_user_id += 1
        user_id = str(self.next_user_id)
        self.users[username] = self.user_factory.create_user(
            username,
            password,
            email,
            user_id,
            role,
        )
        return self.users[username]

    def login_user(self, username, password) -> User:
        if username not in self.users:
            raise ValueError("User does not exist")
        if self.users[username].password != password:
            raise ValueError("Incorrect password")
        return self.users[username]

    def find_user(self, username) -> User | None:
        return self.users.get(username, None)

    def delete_user(self, username) -> bool:
        if username in self.users:
            del self.users[username]
            return True
        else:
            raise ValueError("User does not exist")

    def find_users_by_role(self, role) -> list[User]:
        return [
            user
            for user in self.users.values()
            if user.role.lower() == role.lower()
        ]
